walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,100 @@
+      return Long.compare(a.getWhen().start(), b.getWhen().start());
+    }
+  };
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) { 
+      Collection<TimeRange> expected = Arrays.asList();
+      return expected;
+    } else if (events.isEmpty() || request.getAttendees().isEmpty()) {
+      Collection<TimeRange> expected = Arrays.asList(TimeRange.WHOLE_DAY);
+      return expected;
+    } else {
+      // sort events by start time
+      List<Event> eventsList = new ArrayList<Event>();
+      for (Event event : events) {
+        if (request.getAttendees().containsAll(event.getAttendees())) {
+          eventsList.add(event);
+        }
+      }
+      Collections.sort(eventsList, SORT_BY_START);
+      
+      //ignore people not attending
+      if (eventsList.isEmpty()) {
+        Collection<TimeRange> expected = Arrays.asList(TimeRange.WHOLE_DAY);
+        return expected;
+      }
+
+      //merge event times
+      List<TimeRange> mergedEventsList = new ArrayList<TimeRange>();
+      if (eventsList.size() >= 2) {
+        for (int i = 0; i < eventsList.size()-1; i++) {
+          Event event1 = eventsList.get(i);
+          Event event2 = eventsList.get(i+1);
+          if (event1.getWhen().contains(event2.getWhen())) {
+            mergedEventsList.add(event1.getWhen());
+          } else if (event1.getWhen().overlaps(event2.getWhen())) {
+            mergedEventsList.add(TimeRange.fromStartEnd(event1.getWhen().start(), event2.getWhen().end(), false));
+          } else {
+            mergedEventsList.add(event1.getWhen());
+            mergedEventsList.add(event2.getWhen());
+          }
+        }
+      } else {
+        mergedEventsList.add(eventsList.get(0).getWhen());
+      }
+      
+      //divide time into before event & after event
+      Collection<TimeRange> expected = new ArrayList<TimeRange>();
+      int start = 0, end = 0;
+      if (mergedEventsList.size() >= 2) {
+        if (mergedEventsList.get(0).start() != TimeRange.START_OF_DAY) {
+          expected.add(TimeRange.fromStartEnd(start, mergedEventsList.get(0).start(), false));
+        }
+        for (int i = 0; i < mergedEventsList.size()-1; i++) {
+          TimeRange event1 = mergedEventsList.get(i);
+          TimeRange event2 = mergedEventsList.get(i+1);
+          TimeRange secondHalf = TimeRange.fromStartEnd(event1.end(), event2.start(), false);
+          if (event2.start() - event1.end() >= request.getDuration()) {
+            expected.add(secondHalf);
+          }
+          start = event1.end();
+        }
+
+        if (mergedEventsList.get(mergedEventsList.size()-1).end() < TimeRange.END_OF_DAY) {
+          expected.add(TimeRange.fromStartEnd(mergedEventsList.get(mergedEventsList.size()-1).end(), TimeRange.END_OF_DAY, true));
+        }
+      } else {
```

Comment:
Can we re-order this function so that most of the logic isn't nested in the `if` block? By now, I've forgotten what must be true to have ended up in this `else`.

Reply:
of course

Reply:
https://github.com/gannerr/step/blob/b6c136f564e4ab96a9a02fefec07fd4101d635dc/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java#L88-L111